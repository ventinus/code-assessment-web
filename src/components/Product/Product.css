.product {
  border-radius: 5px;
  background-color: white;
  overflow: hidden; }
  @media (min-width: 768px) {
    .product {
      display: flex;
      flex-direction: row;
      align-items: flex-start; } }

.product__img {
  width: 100%; }
  @media (min-width: 768px) {
    .product__img {
      width: 40%; } }

.product__body {
  padding: 20px; }
  @media (min-width: 768px) {
    .product__body {
      display: flex;
      align-self: stretch;
      flex-direction: column;
      align-items: flex-start;
      width: 60%;
      padding: 30px 35px; } }
  @media (min-width: 1024px) {
    .product__body {
      padding: 40px 50px; } }

.product__body__info {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 100%; }

.product__body__inventory {
  margin-top: 8px;
  text-transform: uppercase; }
  @media (min-width: 1024px) {
    .product__body__inventory {
      margin-top: 16px; } }

.product__body__btn {
  margin-top: 25px; }
  @media (min-width: 768px) {
    .product__body__btn {
      margin-top: auto; } }
