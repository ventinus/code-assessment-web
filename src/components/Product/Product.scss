@import '../../scss/variables/mixins.scss';
@import '../../scss/variables/variables.scss';

.product {
  border-radius: 5px;
  background-color: $white;
  overflow: hidden;

  @include breakpoint(md) {
    display: flex;
    flex-direction: row;
    align-items: flex-start;
  }
}

.product__img {
  width: 100%;

  @include breakpoint(md) {
    width: 40%;
  }
}

.product__body {
  padding: 20px;

  @include breakpoint(md) {
    display: flex;
    align-self: stretch;
    flex-direction: column;
    align-items: flex-start;
    width: 60%;
    padding: 30px 35px;
  }

  @include breakpoint(lg) {
    padding: 40px 50px;
  }
}

.product__body__info {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 100%;
}

.product__body__inventory {
  margin-top: 8px;
  text-transform: uppercase;

  @include breakpoint(lg) {
    margin-top: 16px;
  }
}

.product__body__btn {
  margin-top: 25px;
  @include breakpoint(md) {
    margin-top: auto;
  }
}
