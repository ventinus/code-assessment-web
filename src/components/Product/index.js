import React from 'react'
import PropTypes from 'prop-types'

import {
  PrimaryButton,
} from '..'

import './Product.css'

const imageSrc = file => require(`../../assets/images/${file}`)

const Product = ({ price, inventory, title, src, onAddToCartClicked }) => (
  <article className="product">
    <img src={imageSrc(src)} alt="" className="product__img"/>
    <div className="product__body">
      <div className="product__body__info">
        <h2 className="type--a3">{title}</h2>
        <p className="type--a4">&#36;{price}</p>
      </div>
      <p className="product__body__inventory type--a9 type--grey-6">{inventory || 'None'} remaining</p>

      <PrimaryButton className="product__body__btn" onClick={onAddToCartClicked} disabled={!inventory}>
        {inventory > 0 ? 'Add to cart' : 'Sold Out'}
      </PrimaryButton>
    </div>
  </article>
)

Product.propTypes = {
  price: PropTypes.number,
  inventory: PropTypes.number,
  title: PropTypes.string
}

export default Product
