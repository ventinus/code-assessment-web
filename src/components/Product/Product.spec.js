import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import Product from '.'

const productFactory = (
  price = 9.99,
  inventory = 6,
  title = 'Test Product',
) => ({
  price,
  inventory,
  title,
})

const setup = (props = productFactory()) => {
  const actions = {
    onAddToCartClicked: jest.fn(),
  }

  const component = mount(
    <Product {...props} {...actions} />
  )

  return {
    actions,
    component: component,
    button: component.find('button'),
    inventory: component.find('.product__body__inventory'),
  }
}

describe('Product component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Product {...productFactory()} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('when given inventory', () => {
    it('should render inventory count', () => {
      const { inventory } = setup()
      expect(inventory.text()).toMatch(/^6 remaining/)
    })

    it('should trigger onAddToCartClicked', () => {
      const { button, actions } = setup()
      button.simulate('click')
      expect(actions.onAddToCartClicked).toBeCalled()
    })

    it('should render "Add to cart" in button', () => {
      const { button } = setup()
      expect(button.text()).toMatch(/^Add to cart/)
    })
  })

  describe('when given no inventory', () => {
    it('should not trigger onAddToCartClicked', () => {
      const { button, actions } = setup(productFactory(10, 0))
      button.simulate('click')
      expect(actions.onAddToCartClicked).not.toBeCalled()
    })

    it('should render "Sold Out" in button', () => {
      const { button } = setup(productFactory(10, 0))
      expect(button.text()).toMatch(/^Sold Out/)
    })

    it('should render no inventory', () => {
      const { inventory } = setup(productFactory(10, 0))
      expect(inventory.text()).toMatch(/^None remaining/)
    })
  })
})
