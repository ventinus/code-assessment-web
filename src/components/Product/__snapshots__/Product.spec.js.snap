// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component renders correctly 1`] = `
<article
  className="product"
>
  <img
    className="product__img"
  />
  <div
    className="product__body"
  >
    <div
      className="product__body__info"
    >
      <h2
        className="type--a3"
      >
        Test Product
      </h2>
      <p
        className="type--a4"
      >
        $9.99
      </p>
    </div>
    <p
      className="product__body__inventory type--a9 type--grey-6"
    >
      6
       remaining
    </p>
    <button
      className="button type--a8  button--primary product__body__btn"
      disabled=""
      onClick={[Function]}
      type="button"
    >
      Add to cart
    </button>
  </div>
</article>
`;
